src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs

```diff
@@ -182,18 +183,42 @@ public void SetInput(byte[] inputBuffer, int startIndex, int count)
             Debug.Assert(startIndex >= 0 && count >= 0 && count + startIndex <= inputBuffer.Length);
             Debug.Assert(!_inputBufferHandle.IsAllocated);
 
-            if (0 == count)
+            SetInput(inputBuffer.AsMemory().Slice(startIndex, count));
+        }
+
+        public void SetInput(ReadOnlyMemory<byte> inputBuffer)
+        {
+            Debug.Assert(NeedsInput(), "We have something left in previous input!");
+            Debug.Assert(!_inputBufferHandle.IsAllocated);
+
+            if (inputBuffer.Length == 0)
                 return;
 
             lock (SyncLock)
             {
-                _inputBufferHandle = GCHandle.Alloc(inputBuffer, GCHandleType.Pinned);
-                _zlibStream.NextIn = _inputBufferHandle.AddrOfPinnedObject() + startIndex;
-                _zlibStream.AvailIn = (uint)count;
+                _inputBufferHandle = new InputBufferHandle(inputBuffer.Pin());
+                _zlibStream.NextIn = _inputBufferHandle.Pointer;
+                _zlibStream.AvailIn = (uint)inputBuffer.Length;
                 _finished = false;
             }
         }
 
+        private struct InputBufferHandle
```

Comment:
I don't see why this new struct is needed (was it just to give these operations the same name as GCHandle to avoid additional changes elsewhere?).  _inputBufferHandle should just be typed as MemoryHandle.

Reply:
I made the struct to avoid having to mark more methods as `unsafe`. I can remove it if that's not a problem.

Reply:
It's not a problem to surround the regions that use pointers with unsafe (the whole method needn't be marked if that would significantly increase the scope).  Whether unsafe is there or not, a pointer is still being used; there's actually a benefit in "unsafe" showing up where the pointer is being used, as it makes it clear that something potentially dangerous is happening.

Reply:
Fixed.